docs: add extended passport support (#142)
decentraland/adr: content/ADR-138-extended-passport-support.md

---

> adr: 138

Reviewer: Please use ADR id same as PR (in this case 143) to avoid conflicts with other ADRs (there's already another one ongoing with same id).

---
adr: 138
date: 2022-11-25
title: Technical Assessment for the Extended Passport
authors:
  - manumena
status: Draft
type: RFC
spdx-license: CC0-1.0
redirect_from:
  - /rfc/RFC-1
---

## Need

A New Passport view is being developed by the Explorer.

> This view will require storing information in the profile entity and retrieving things as collectibles to be displayed, as well as being able to block, add as a friend or initiate a chat.

Reviewer: ```suggestion
This view requires storing some information in the profile entity and retrieving other data (such as collectibles) to be displayed. It also allows blocking users, adding a user as a friend or initiating a chat.
```

## Requirements
The applicable requirements from the Catalyst point of view:

1. Intro of a user:
Use the "description" field from profiles to use in the "Intro" section or create a new one if it is different from the "description".

2. Block/Unblock users:
Use the "blocked" field from profiles to see/report blocked users.

3. Equipped wearables:
Take the information from the profile. There are the equipped wearables, emotes and the current name.

4. Items: wearables/emotes/names/lands:

> For the wearables and emotes there are two available endpoints `/wearables-by-owner` and `/emotes-by-owner` that can be used for that. It will be necessary to create `/names-by-owner` and `/lands-by-owner` endpoints with similar behavior.

Reviewer: Let's try to define whether we'll be needing pagination or not. If we do, I'd probably create new endpoints for all, as the existing ones will require breaking changes. If no pagination is needed, then we're good to implement the newer ones just the same.

Author: 👍🏻

> 5. Guest: Passport for guest users will show only name and description, there is nothing to do for guests from the catalyst end.

Reviewer: ```suggestion
5. Guest: Passport for guest users will show only name and description, there is nothing to do for guests from the Catalyst end.
```

## Performance

> To load a passport, first, there will be made 5 requests to a Catalyst (lambdas):

Reviewer: ```suggestion
To load a passport, 5 requests to a Catalyst (lambdas) need to be made:
```

- `/profiles`
- `/wearables-by-owner`
- `/emotes-by-owner`
- `/lands-by-owner`
- `/names-by-owner`

> Once this information is obtained, to display wearables and emotes it will be needed a couple of requests per each emote/wearable to download its image/thumbnail to be shown. This does not scale because some users have more than 1k NFTs. So the explorer should implement a lazy loading mechanism (for image/thumbnail) as the user browses the collectibles.

Reviewer: ```suggestion
Once this information is obtained, to display wearables and emotes a few extra requests will be needed for each emote/wearable to download their image/thumbnail. This does not scale because some users have more than 1k NFTs. So the Explorer should implement a lazy loading mechanism (for image/thumbnail) as the user browses the collectibles.
```

> All except /profiles are expensive requests. Internally, from 2 to 5 they make queries to The Graph to check owned NFTs urns and then to the content server to get the definitions. Currently, 2 and 3 are not paginated, so making a query for a user that owns more than 1k wearables would be considerably heavy. Also, sometimes queries to The Graph are slow. So to improve this situation the endpoints from 2 to 5 could be paginated.

Reviewer: ```suggestion
All endpoints except `/profiles` are expensive requests. Internally, endpoints 2 to 5 make queries to TheGraph to check owned NFTs urns and then to the Content Server to get the definitions. Currently, endpoints 2 and 3 are not paginated, so making a query for a user that owns more than 1k wearables would be considerably heavy. Also, sometimes queries to TheGraph are slow. So to improve this situation the endpoints from 2 to 5 should be paginated.
```

## Endpoints details

## Estimate

### MVP
The requirements are not blocking to start the client-side development.

Implement `/lands-by-owner` and `/names-by-owner` 

### Performance Improvements

> - Add pagination to the existing endpoints

Reviewer: 🤔 This would break current APIs. If we need pagination I'd go for a fresh set of endpoints in Lambdos.

Author: I agree
